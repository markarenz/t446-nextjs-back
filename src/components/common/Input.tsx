import styles from '../../styles/modules/Input.module.scss';

type Props = {
  name?: string;
  type?: string;
  value?: string;
  onChange: Function;
  onEnterkey?: Function;
  tabIndex?: number;
  autoFocus: boolean;
  rows?: number;
  maxLen?: number;
  testId?: string;
};
type OptionalInputProps = {
  maxLength?: number;
};
const Input: React.FC<Props> = ({
  name,
  type,
  value,
  onChange,
  onEnterkey,
  tabIndex,
  autoFocus,
  rows,
  maxLen,
  testId
}) => {
  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter') {
      if (typeof onEnterkey === 'function') {
        onEnterkey();
      }
    }
  };
  const optionalProps: OptionalInputProps = {};
  if (maxLen) {
    optionalProps.maxLength = maxLen;
  }
  if (rows && rows > 1) {
    return (
      <textarea
        rows={rows}
        name={name}
        value={value}
        onChange={(e) => onChange(e)}
        tabIndex={tabIndex}
        autoFocus={autoFocus}
        className={styles.textarea}
        data-testid={testId}
        {...optionalProps}
      />
    );
  }
  return (
    <input
      name={name}
      type={type}
      value={value}
      onChange={(e) => onChange(e)}
      tabIndex={tabIndex}
      autoFocus={autoFocus}
      onKeyDown={handleKeyDown}
      className={styles.input}
      data-testid={testId}
      {...optionalProps}
    />
  );
};

export default Input;
