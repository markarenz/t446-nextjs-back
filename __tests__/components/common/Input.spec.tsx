import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Input from '../../../src/components/common/Input';

const defaultProps = {
  name: 'test',
  type: 'text',
  value: 'foo',
  onChange: jest.fn(),
  onEnterkey: jest.fn(),
  tabIndex: 1,
  autoFocus: false,
  maxLen: 100,
  testId: 'test-input'
};

const textAreaProps = {
  ...defaultProps,
  rows: 4
};

const noEnterKeyProps = {
  ...defaultProps,
  onEnterKey: null
};
describe('Input', () => {
  it('renders component', async () => {
    render(<Input {...defaultProps} />);
    const element = await screen.findByTestId('test-input');
    expect(element).toBeInTheDocument();
  });
  it('handles onChange', async () => {
    render(<Input {...defaultProps} />);
    const element = await screen.findByTestId('test-input');
    fireEvent.change(element, { target: { value: 'new' } });
    expect(defaultProps.onChange).toHaveBeenCalled;
  });
  it('renders textarea with change', async () => {
    render(<Input {...textAreaProps} />);
    const element = await screen.findByTestId('test-input');
    fireEvent.change(element, { target: { value: 'new' } });
    expect(defaultProps.onChange).toHaveBeenCalled;
  });

  it('handles keyDown enter key', async () => {
    render(<Input {...defaultProps} />);
    const element = await screen.findByTestId('test-input');
    fireEvent.keyDown(element, { key: 'Enter', code: 13, charCode: 13 });
    expect(defaultProps.onEnterkey).toHaveBeenCalled;
  });
  it('handles keyDown non-enter key', async () => {
    render(<Input {...noEnterKeyProps} />);
    const element = await screen.findByTestId('test-input');
    fireEvent.keyDown(element, { key: 'A', code: 65, charCode: 65 });
    expect(defaultProps.onEnterkey).not.toHaveBeenCalled;
  });
});
