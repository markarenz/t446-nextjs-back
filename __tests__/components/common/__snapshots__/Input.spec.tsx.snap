// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders component 1`] = `
<input
  class="input"
  data-testid="test-input"
  maxlength="100"
  name="test"
  tabindex="1"
  type="text"
  value="foo"
/>
`;
